CenterForOpenScience/osf.io
# fix: less surprising RelationshipField

## Purpose
in some cases, `api.base.serializers.RelationshipField` requires that the serializer field have the same name as the corresponding model field, which is a problematic entanglement of the api and data layers.

(encountered/fixed while working on #10273  )


## Changes
digging to fix this bug led to a broader cleanup:
- since `RelationshipField` overrides all meaningful logic in `HyperlinkedIdentityField` (with some extra logic to counteract it), change to simply extend `Field`.
- extending `Field` removes an implicit `source='*'`, which is good because RelationshipField subclasses with `to_internal_value` don't always have to know the name they're bound to on the serializer.
- update `FilterMixin` to:
  - remove unnecessary `parse_sparse_fields` call, which allows us to
  - use bound serializer fields (which have `source` correctly set).
  - prefer `filter_key`, when it exists.
- remove unused sparse-fieldset logic and tidy the rest a bit.
- consolidate some redundant utils (`api.base.utils.decompose_field`, `JSONAPISerializer.get_unwrapped_field').



## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify relationships in the api display, embed, and filter as expected.

What are the areas of risk?

Any concerns/considerations/questions that development raised?

## Documentation



## Side Effects



## Ticket
